### DISCUSSION TITLE
RFC: Rspack v1.0 Release Plan

### DISCUSSION THREAD

**Discussion by chenjiahan** [web-infra-dev/rspack#6315] [Release Plan]
![Rspack 1 0 planning](https://github.com/web-infra-dev/rspack/assets/7237365/d420229d-4a5e-42ab-94ec-c03303cf11a3)

# Introduction

We are excited to announce the planned release of Rspack v1.0 in July 2024!

This release will include several key features aimed at improving the functionality and performance of Rspack. Below is an overview of the key features ahead of the release.

# Key Features Before v1.0

## 1. Lazy Compilation Support

Lazy compilation will significantly improve the performance of large applications that utilize dynamic imports or multiple entries. This feature is particularly crucial for static site generators like [Rspress](https://github.com/web-infra-dev/rspress). 

Imagine some code that you might not use on the first run, for example, when you click a button, another page loads, if you don't click that button at all when you're developing other stuff, that page is useless, so Rspack don't need to build it. Rspack only builds it when we do the click.

> Progress on this feature can be tracked on [Lazy Compilation Pull Request](https://github.com/web-infra-dev/rspack/pull/5915).

## 2. Enhanced Tree Shaking

We have implemented tree shaking based on esbuild's strategy. While effective in many scenarios, it does face limitations affecting optimizations such as barrel optimization and circular dependencies. From v0.4.2, we've introduced new tree shaking algorithms from webpack, which is enabled by default starting from v0.6.0.

> See [Announcing Rspack v0.6](https://www.rspack.dev/blog/announcing-0.6#enable-new-tree-shaking-by-default) for more details.

## 3. Native CSS Extract Plugin Support

Our experiments css support have revealed some limitations, particularly in aligning with webpack's handling of legacy browsers and not working with css-loader. To address this and ease the migration for applications using css-loader, we've introduced a native CSS extract plugin that mirrors the behavior of webpack's mini-css-extract-plugin. This plugin, which supports legacy browsers and css-loader effectively, has been included since version 0.6.

## 4. ModuleConcatenationPlugin Support

The ModuleConcatenationPlugin facilitates "scope hoisting", which can reduce bundle code size and enhance execution speed by minimizing the number of `webpack_require` calls. This feature is foundational for high-quality ESM output. We are collaborating with the webpack team to refine the ESM output quality of both Rspack and webpack, laying the groundwork for Rslib.

# Architecture

## Stabilizing the Public API

After several iterations from versions 0.1 to 0.5, most APIs have stabilized. However, certain APIs still require refinement before v1.0:

- `Configuration`: We will remove all deprecated configurations, stabilizing all configurations except those under rspackFuture.
- `Hooks API`: We support a subset of webpack hooks and will stabilize the officially supported hooks.
- `Compiler & Compilation objects`: While we do not plan to support all properties and methods of webpack's compiler and compilation objects, we will stabilize the supported methods and properties.
- `JavaScript interface`: Rspack will offer a JavaScript API that supports custom build or development processes, allowing better framework integration.

## Website Redesign

Our website will undergo a redesign to reflect the numerous feature updates and architectural overhauls since v0.1. The redesign will include:

- Updated information on supported webpack loaders and plugins.
- Rewritten guides and configuration sections to reflect the latest features of v1.0.
- A new API page detailing stabilized public APIs.
- A contribution guide for writing Rust or JavaScript plugins and loaders.
- An FAQ section addressing common queries.

## Build Artifacts Improvement

We aim to align our build artifacts more closely with webpack, addressing inconsistencies like CSS extraction and interop runtime. Our goal is to achieve 100% alignment with webpack build output by v1.0.

# What’s Next

The release of v1.0 is just the beginning for Rspack. We have big plans for its future development, including:

- `Improved performance`: We will continue enhancing Rspack's architecture to boost build performance significantly.
- `First-class ESM output support`: We aim to produce high-quality ESM outputs that do not compromise treeshaking efficiency.
- `Portable persistent cache support`: To facilitate large-scale projects, we will explore portable persistent caching to distribute the build process across different machines.
- `Stable Rust API`: To integrate more efficiently with frameworks, we will develop a stable Rust API.
- `React Server Components support`: We are exploring ways to natively support React Server Components.
- `WebAssembly compatible`: Compiling Rspack to WebAssembly will enable it to run in browser environments, suitable for online web playgrounds or platforms like StackBlitz.
- `Native plugin support`: We plan to research higher-performance plugin solutions beyond the current Wasm plugin.
We welcome feedback on these initiatives to ensure they meet the needs of our users.

**Comment by AsuraKev** [comment]
does v1 also include Module federation stable version?

**Reply by chenjiahan** [reply]
Do you mean Module Federation 2.0? It will be a standalone plugin and we have no plans to builtin it into Rspack yet.

**Reply by AsuraKev** [reply]
I see, will that also come out of experimental in July?

**Reply by chenjiahan** [reply]
Module Federation 2.0 is stable now and we will release the Rspack v1.0 in July

**Comment by dxhsmwqcwelllee** [comment]
v1 can include Persistent Cache ?

**Reply by chenjiahan** [reply]
Rspack v1.0 will not include the persistent cache feature, it will be implemented in future iterations.

**Reply by Zavieer** [reply]
> Rspack v1.0 will not include the persistent cache feature, it will be implemented in future iterations.

This feature seems quite important for performance, right? I'd like to know if it hasn't been implemented because of low priority or because of the large amount of work involved. 
Thanks again to the rspack team for their efforts

**Reply by chenjiahan** [reply]
Persistent cache is important and has a high priority, but it requires a lot of work 😄

**Reply by dxhsmwqcwelllee** [reply]
> Rspack v1.0 will not include the persistent cache feature, it will be implemented in future iterations.

Is there an approximate time ？

**Reply by chenjiahan** [reply]
@dxhsmwqcwelllee Should be Q4 2024

**Reply by dxhsmwqcwelllee** [reply]
<img width="990" alt="image" src="https://github.com/user-attachments/assets/4dabdd2e-6ca5-46f6-862a-5008d2c72f26" />

The documents have been released, how long is the 1.2.0 version released?  0_0

**Comment by lzxb** [comment]
I plan to implement an SSR rendering framework based on Rpack and module federation + ESM

**Comment by rafaell-lycan** [comment]
Will v1.0 bring better/extended support to Webpack plugins?

I plan to try `rspack` at @Typeform for a week on our current build pipeline (custom webpack + esbuild) and depending on the results I might suggest a change, so I am waiting for the v1.0 release to see if this a great alternative for us and perhaps discard Vite migration from the table as deploying ~20 micro-frontends that uses webpack module federation will probably break a lot of things. 😅 

We have a set of plugins that are crucial for us:
- ~esbuild-loader~ 💀
- PostCSS & Autoprefixer ✅
- [mini-css-extract-plugin](https://www.npmjs.com/package/mini-css-extract-plugin) ✅
- [babel-plugin-styled-components](https://www.npmjs.com/package/babel-plugin-styled-components) ✅
- [@pmmmwh/react-refresh-webpack-plugin](https://www.npmjs.com/package/@pmmmwh/react-refresh-webpack-plugin) ⚠️
- [svg-sprite-loader](https://www.npmjs.com/package/svg-sprite-loader) ❓

We also have some custom Webpack and Node.js plugins for metrics & bundle size which I can manage to convert later on.

**Reply by chenjiahan** [reply]
We have been continuously improving Rspack's compatibility with webpack plugins. See https://rspack.dev/guide/compatibility/plugin for plugin compatibility

**Comment by Sintayew4** [comment]
#315

**Comment by dartess** [comment]
Hey. Do you have plans to update dependencies to the first version?

npm audit:
```
ws  8.0.0 - 8.17.0
Severity: high
ws affected by a DoS when handling a request with many HTTP headers - https://github.com/advisories/GHSA-3h5v-q93c-6h6q
fix available via `npm audit fix --force`
Will install @rspack/cli@0.0.6, which is a breaking change
node_modules/ws
  @rspack/dev-server  <=0.0.0-ff9366411-20230221070252 || >=0.0.8
  Depends on vulnerable versions of ws
  node_modules/@rspack/dev-server
    @rspack/cli  <=0.0.0-ff9366411-20230221070252 || >=0.0.8
    Depends on vulnerable versions of @rspack/dev-server
    node_modules/@rspack/cli

```

also warnings during installation:

```
npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
npm WARN deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
```
via `@rspack/dev-server` -> `webpack-dev-server@4.13.1`

**Reply by hardfist** [reply]
we will upgrade webpack-dev-server to latest version

**Comment by karthickshanmugam0689** [comment]
Does this discussion relate to https://github.com/web-infra-dev/rspack/releases/tag/v1.0.0-alpha.2 ?

**Comment by nailfar** [comment]
物理缓存什么时候做，项目太大几个热更新内存崩了，有时候做缝合怪项目，要把四个项目合一，四个项目同开根本没法玩

**Reply by chenjiahan** [reply]
Work in progress

**Comment by imzisy** [comment]
Is there any support for PWA? If not, is it planned for a future release? @chenjiahan

**Reply by chenjiahan** [reply]
There are some community plugins for PWA support, see https://github.com/web-infra-dev/awesome-rspack?tab=readme-ov-file#rspack-plugins

**Reply by hardfist** [reply]
the workbox-webpack-plugin is already official supported

**Comment by hasnum-stack** [comment]
你好, rslib估计什么时候开源,是否支持bundless模式

**Reply by chenjiahan** [reply]
Rslib repository is now public and it will support bundleless output: https://github.com/web-infra-dev/rslib

**Comment by famicomp** [comment]
Is there an updated timeline on when Rspack v1.0 is planned to be released?

**Reply by chenjiahan** [reply]
We are doing the final review and writing the release blog, it will take 2-3 weeks 😄
